File: apps/backend/app/schemas/analysis.py
"""
Analysis API schemas.
"""

from __future__ import annotations

from datetime import datetime

from app.models.analysis import AnalysisStatus
from pydantic import BaseModel, ConfigDict, Field


class AnalysisResponse(BaseModel):
    """
    API representation of an analysis job.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    repository_id: int
    status: AnalysisStatus
    progress: int = Field(ge=0, le=100)
    started_at: datetime | None = None
    completed_at: datetime | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class AnalysisListResponse(BaseModel):
    """
    Paginated analysis list response.
    """

    items: list[AnalysisResponse]
    total: int
    page: int
    per_page: int
    pages: int


class RepositoryMetricsResponse(BaseModel):
    """
    Repository metrics produced by an analysis.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    analysis_result_id: int
    loc: int
    files: int
    classes: int
    functions: int
    complexity: float | None = None
    maintainability: float | None = None


class DetectedTechnologyResponse(BaseModel):
    """
    Technology detected during analysis.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    analysis_result_id: int
    technology: str
    version: str | None = None
    confidence_score: float


class DependencyGraphResponse(BaseModel):
    """
    Dependency graph produced by an analysis.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    analysis_result_id: int
    graph_data: str
    created_at: datetime


class AnalysisResultResponse(BaseModel):
    """
    Complete analysis result.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    analysis_job_id: int
    summary: str | None = None
    created_at: datetime
    metrics: RepositoryMetricsResponse | None = None
    technologies: list[DetectedTechnologyResponse]
    dependency_graph: DependencyGraphResponse | None = None

File: apps/frontend/src/types/analysis.ts
export type AnalysisStatus = "pending" | "queued" | "running" | "completed" | "failed" | "cancelled";

export interface Analysis {
  id: number;
  repository_id: number;
  status: AnalysisStatus;
  progress: number;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface AnalysisListResponse {
  items: Analysis[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface AnalysisMetrics {
  id: number;
  analysis_result_id: number;
  loc: number;
  files: number;
  classes: number;
  functions: number;
  complexity: number;
  maintainability: number;
}

export interface AnalysisTechnology {
  id: number;
  analysis_result_id: number;
  technology: string;
  version: string | null;
  confidence_score: number;
}

export interface AnalysisDependencyGraph {
  id: number;
  analysis_result_id: number;
  graph_data: string;
  created_at: string;
}

export interface AnalysisResult {
  id: number;
  analysis_job_id: number;
  summary: string | null;
  created_at: string;
  metrics: AnalysisMetrics;
  technologies: AnalysisTechnology[];
  dependency_graph: AnalysisDependencyGraph | null;
}